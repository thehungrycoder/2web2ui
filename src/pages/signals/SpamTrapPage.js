import React, { Component, Fragment } from 'react';
import { Link } from 'react-router-dom';
import { Panel, Grid } from '@sparkpost/matchbox';
import Page from './components/SignalsPage';
import BarChart from './components/charts/barchart/BarChart';
import SpamTrapActions from './components/actionContent/SpamTrapActions';
import TooltipMetric from './components/charts/tooltip/TooltipMetric';
import DateFilter from './components/filters/DateFilter';
import { SPAM_TRAP_INFO } from './constants/info';
import withSpamTrapDetails from './containers/SpamTrapDetailsContainer';
import { Loading } from 'src/components';
import Callout from 'src/components/callout';
import OtherChartsHeader from './components/OtherChartsHeader';
import Calculation from './components/viewControls/Calculation';
import ChartHeader from './components/ChartHeader';
import { formatFullNumber, formatNumber, roundToPlaces } from 'src/helpers/units';
import moment from 'moment';
import _ from 'lodash';

import EngagementRecencyPreview from './components/previews/EngagementRecencyPreview';
import HealthScorePreview from './components/previews/HealthScorePreview';
import styles from './DetailsPages.module.scss';

export class SpamTrapPage extends Component {
  state = {
    selectedDate: null,
    calculation: 'relative'
  }

  componentDidMount() {
    const { selected } = this.props;

    if (selected) {
      this.setState({ selectedDate: selected });
    }
  }

  componentDidUpdate(prevProps) {
    const { data } = this.props;
    const { selectedDate } = this.state;

    const dataSetChanged = prevProps.data !== data;
    let selectedDataByDay = _.find(data, ['date', selectedDate]);

    // Select last date in time series
    if (dataSetChanged && !selectedDataByDay) {
      selectedDataByDay = _.last(data);
      this.setState({ selectedDate: selectedDataByDay.date });
    }
  }

  handleCalculationToggle = (value) => {
    this.setState({ calculation: value });
  }

  handleDateSelect = (node) => {
    this.setState({ selectedDate: _.get(node, 'payload.date') });
  }

  getYAxisProps = () => {
    const { data } = this.props;
    const { calculation } = this.state;

    return {
      tickFormatter: calculation === 'relative' ? (tick) => `${roundToPlaces(tick * 100, 2)}%` : (tick) => formatNumber(tick),
      domain: data.every(({ relative_trap_hits }) => !relative_trap_hits) && calculation === 'relative'
        ? [0, 1] : ['auto', 'auto']
    };
  }

  getXAxisProps = () => {
    const { xTicks } = this.props;
    return {
      ticks: xTicks,
      tickFormatter: (tick) => moment(tick).format('M/D')
    };
  }

  getTooltipContent = ({ payload = {}}) => (
    <Fragment>
      <TooltipMetric label='Spam Trap Hits' value={formatFullNumber(payload.trap_hits)} />
      <TooltipMetric label='Injections' value={formatFullNumber(payload.injections)} />
      <TooltipMetric label='Spam Trap Rate' value={`${roundToPlaces(payload.relative_trap_hits * 100, 4)}%`} />
    </Fragment>
  )

  renderContent = () => {
    const { data = [], loading, gap, empty, error } = this.props;
    const { calculation, selectedDate } = this.state;
    const selectedData = _.find(data, ['date', selectedDate]) || {};
    let chartPanel;

    if (empty) {
      chartPanel = <Callout title='No Data Available'>Insufficient data to populate this chart</Callout>;
    }

    if (error) {
      chartPanel = <Callout title='Unable to Load Data'>{error.message}</Callout>;
    }

    if (loading) {
      chartPanel = (
        <div style={{ height: '220px', position: 'relative' }}>
          <Loading />
        </div>
      );
    }

    return (
      <Grid>
        <Grid.Column sm={12} md={7}>
          <Panel sectioned>
            <ChartHeader
              title='Spam Trap Monitoring'
              primaryArea={
                <Calculation
                  initialSelected={calculation}
                  onChange={this.handleCalculationToggle}
                />
              }
              tooltipContent={SPAM_TRAP_INFO}
            />
            {chartPanel || (
              <BarChart
                gap={gap}
                onClick={this.handleDateSelect}
                selected={selectedDate}
                timeSeries={data}
                tooltipContent={this.getTooltipContent}
                yKey={calculation === 'absolute' ? 'trap_hits' : 'relative_trap_hits'}
                yAxisProps={this.getYAxisProps()}
                xAxisProps={this.getXAxisProps()}
              />
            )}
          </Panel>
        </Grid.Column>
        <Grid.Column sm={12} md={5} mdOffset={0}>
          <div className={styles.OffsetCol}>
            {!chartPanel && <SpamTrapActions percent={selectedData.relative_trap_hits} date={selectedDate} />}
          </div>
        </Grid.Column>
      </Grid>
    );
  }

  render() {
    const { facet, facetId } = this.props;

    return (
      <Page
        breadcrumbAction={{ content: 'Back to Overview', to: '/signals', component: Link }}
        dimensionPrefix='Spam Trap Monitoring for'
        facet={facet}
        facetId={facetId}
        primaryArea={<DateFilter />}>
        {this.renderContent()}
        <OtherChartsHeader facet={facet} facetId={facetId} />
        <Grid>
          <Grid.Column xs={12} sm={6}>
            <EngagementRecencyPreview />
          </Grid.Column>
          <Grid.Column xs={12} sm={6}>
            <HealthScorePreview />
          </Grid.Column>
        </Grid>
      </Page>
    );
  }
}

export default withSpamTrapDetails(SpamTrapPage);
