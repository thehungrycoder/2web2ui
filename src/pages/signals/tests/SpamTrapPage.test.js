import { shallow } from 'enzyme';
import React from 'react';
import { SpamTrapPage } from '../SpamTrapPage';

describe('Signals Spam Trap Page', () => {
  let wrapper;
  let props;
  const data = [
    {
      date: '2017-01-01',
      relative_trap_hits: 0.1
    },
    {
      date: '2017-01-02',
      relative_trap_hits: 0.5
    }
  ];

  beforeEach(() => {
    props = {
      facetId: 'test.com',
      facet: 'sending-domain',
      data: [{ relative_trap_hits: 0.1 }],
      gap: 0.25,
      loading: false,
      empty: false,
      selected: '2017-01-01',
      xTicks: []
    };
    wrapper = shallow(<SpamTrapPage {...props}/>);
    wrapper.setProps({ data });
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders loading correctly', () => {
    wrapper.setProps({ loading: true });
    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders empty correctly', () => {
    wrapper.setProps({ empty: true });
    expect(wrapper.find('Panel')).toMatchSnapshot();
  });

  it('renders error correctly', () => {
    wrapper.setProps({ error: { message: 'error message' }});
    expect(wrapper).toMatchSnapshot();
  });

  describe('local state', () => {
    it('handles calculation type', () => {
      const calculation = shallow(wrapper.find('ChartHeader').props().primaryArea);
      calculation.simulate('change', 'relative');
      expect(wrapper.find('BarChart').prop('yKey')).toEqual('relative_trap_hits');
    });

    it('handles date select', () => {
      wrapper.find('BarChart').simulate('click', { payload: { date: '2017-01-02' }});
      expect(wrapper.find('BarChart').prop('selected')).toEqual('2017-01-02');
      expect(wrapper.find('SpamTrapActions')).toMatchSnapshot();
    });

    it('sets selected date on mount if provided one', () => {
      wrapper = shallow(<SpamTrapPage {...props} selected='initial-selected'/>);
      expect(wrapper.find('BarChart').prop('selected')).toEqual('initial-selected');
      expect(wrapper.find('SpamTrapActions').prop('date')).toEqual('initial-selected');
    });

    it('uses last selected date if selected date is not in data', () => {
      wrapper = shallow(<SpamTrapPage {...props} loading={true} selected='initial-selected'/>);
      wrapper.setProps({ data: [1, { date: 'last-date' }], loading: false });
      expect(wrapper.find('BarChart').prop('selected')).toEqual('last-date');
      expect(wrapper.find('SpamTrapActions').prop('date')).toEqual('last-date');
    });

    it('does not use last selected date if selected date is in data', () => {
      wrapper = shallow(<SpamTrapPage {...props} selected='first-date'/>);
      wrapper.setProps({ data: [{ date: 'first-date' }, { date: 'last-date' }]});
      expect(wrapper.find('BarChart').prop('selected')).toEqual('first-date');
      expect(wrapper.find('SpamTrapActions').prop('date')).toEqual('first-date');
    });
  });

  describe('bar chart props', () => {
    it('renders tooltip content', () => {
      const Tooltip = wrapper.find('BarChart').prop('tooltipContent');
      expect(shallow(<Tooltip payload={{
        trap_hits: 2,
        injections: 3,
        relative_trap_hits: 0.001234567
      }} />)).toMatchSnapshot();
    });

    it('gets x axis props', () => {
      const axisProps = wrapper.find('BarChart').prop('xAxisProps');
      expect(axisProps).toMatchSnapshot();
      expect(axisProps.tickFormatter('2018-12-05')).toEqual('12/5');
    });

    it('gets y axis props with relative calculation', () => {
      wrapper.setState({ calculation: 'relative' });
      const axisProps = wrapper.find('BarChart').prop('yAxisProps');
      expect(axisProps.tickFormatter(0.2468)).toEqual('24.68%');
    });

    it('gets y axis props with absolute calculation', () => {
      wrapper.setState({ calculation: 'absolute' });
      const axisProps = wrapper.find('BarChart').prop('yAxisProps');
      expect(axisProps.tickFormatter((2468))).toEqual('2.47K');
      expect(axisProps.domain).toEqual(['auto', 'auto']);
    });

    it('gets y axis props with relative calculation and 0 values', () => {
      wrapper.setState({ calculation: 'relative' });
      wrapper.setProps({ data: [{ relative_trap_hits: 0 }, { relative_trap_hits: null }]});
      const axisProps = wrapper.find('BarChart').prop('yAxisProps');
      expect(axisProps.domain).toEqual([0, 1]);
    });
  });
});
