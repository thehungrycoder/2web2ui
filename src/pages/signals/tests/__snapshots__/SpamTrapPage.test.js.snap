// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals Spam Trap Page bar chart props gets x axis props 1`] = `
Object {
  "tickFormatter": [Function],
  "ticks": Array [],
}
`;

exports[`Signals Spam Trap Page bar chart props renders tooltip content 1`] = `
<Fragment>
  <TooltipMetric
    label="Spam Trap Hits"
    value="2"
  />
  <TooltipMetric
    label="Injections"
    value="3"
  />
  <TooltipMetric
    label="Spam Trap Rate"
    value="0.1235%"
  />
</Fragment>
`;

exports[`Signals Spam Trap Page local state handles date select 1`] = `
<SpamTrapActions
  date="2017-01-02"
  percent={0.5}
/>
`;

exports[`Signals Spam Trap Page renders correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Overview",
      "to": "/signals",
    }
  }
  dimensionPrefix="Spam Trap Monitoring for"
  facet="sending-domain"
  facetId="test.com"
  primaryArea={<Connect(DateFilter) />}
>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          primaryArea={
            <Calculation
              initialSelected="relative"
              onChange={[Function]}
            />
          }
          title="Spam Trap Monitoring"
          tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
        />
        <BarChart
          fill="#B157CE"
          gap={0.25}
          height={250}
          margin={
            Object {
              "bottom": 5,
              "left": 18,
              "right": 0,
              "top": 12,
            }
          }
          onClick={[Function]}
          selected="2017-01-01"
          timeSeries={
            Array [
              Object {
                "date": "2017-01-01",
                "relative_trap_hits": 0.1,
              },
              Object {
                "date": "2017-01-02",
                "relative_trap_hits": 0.5,
              },
            ]
          }
          tooltipContent={[Function]}
          width="99%"
          xAxisProps={
            Object {
              "tickFormatter": [Function],
              "ticks": Array [],
            }
          }
          xKey="date"
          yAxisProps={
            Object {
              "domain": Array [
                "auto",
                "auto",
              ],
              "tickFormatter": [Function],
            }
          }
          yKey="relative_trap_hits"
          yRange={
            Array [
              "auto",
              "auto",
            ]
          }
        />
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div
        className="OffsetCol"
      >
        <SpamTrapActions
          date="2017-01-01"
          percent={0.1}
        />
      </div>
    </Grid.Column>
  </Grid>
  <OtherChartsHeader
    facet="sending-domain"
    facetId="test.com"
  />
  <Grid>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(withEngagementRecencyDetails(EngagementRecencyPreview))) />
    </Grid.Column>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithHealthScoreDetails(HealthScorePreview))) />
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Spam Trap Page renders empty correctly 1`] = `
<Panel
  sectioned={true}
>
  <ChartHeader
    primaryArea={
      <Calculation
        initialSelected="relative"
        onChange={[Function]}
      />
    }
    title="Spam Trap Monitoring"
    tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
  />
  <Callout
    title="No Data Available"
  >
    Insufficient data to populate this chart
  </Callout>
</Panel>
`;

exports[`Signals Spam Trap Page renders error correctly 1`] = `
<SignalsPage
  breadcrumbAction={
    Object {
      "component": [Function],
      "content": "Back to Overview",
      "to": "/signals",
    }
  }
  dimensionPrefix="Spam Trap Monitoring for"
  facet="sending-domain"
  facetId="test.com"
  primaryArea={<Connect(DateFilter) />}
>
  <Grid>
    <Grid.Column
      md={7}
      sm={12}
    >
      <Panel
        sectioned={true}
      >
        <ChartHeader
          primaryArea={
            <Calculation
              initialSelected="relative"
              onChange={[Function]}
            />
          }
          title="Spam Trap Monitoring"
          tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
        />
        <Callout
          title="Unable to Load Data"
        >
          error message
        </Callout>
      </Panel>
    </Grid.Column>
    <Grid.Column
      md={5}
      mdOffset={0}
      sm={12}
    >
      <div
        className="OffsetCol"
      />
    </Grid.Column>
  </Grid>
  <OtherChartsHeader
    facet="sending-domain"
    facetId="test.com"
  />
  <Grid>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(withEngagementRecencyDetails(EngagementRecencyPreview))) />
    </Grid.Column>
    <Grid.Column
      sm={6}
      xs={12}
    >
      <withRouter(Connect(WithHealthScoreDetails(HealthScorePreview))) />
    </Grid.Column>
  </Grid>
</SignalsPage>
`;

exports[`Signals Spam Trap Page renders loading correctly 1`] = `
<Panel
  sectioned={true}
>
  <ChartHeader
    primaryArea={
      <Calculation
        initialSelected="relative"
        onChange={[Function]}
      />
    }
    title="Spam Trap Monitoring"
    tooltipContent="
  The share over time of your total email that has been sent to known spam traps. An
  excessive rate of spam trap hits is an indicator of poor list procurement and hygiene practices.
"
  />
  <div
    style={
      Object {
        "height": "220px",
        "position": "relative",
      }
    }
  >
    <Loading />
  </div>
</Panel>
`;
